tools/buildgen/extract_metadata_from_bazel_xml.py

```diff
@@ -0,0 +1,709 @@
+        public_name = bazel_label_to_dep_name.get(transitive_bazel_dep, None)
+        if public_name:
+            deps.add(public_name)
+        external_dep_name_maybe = _external_dep_name_from_bazel_dependency(transitive_bazel_dep)
+        if external_dep_name_maybe:
+            # expanding all absl libraries is technically correct but creates too much noise
+            if not external_dep_name_maybe.startswith('absl'):
+                deps.add(external_dep_name_maybe)
+
+    target_dict['public_headers'] = list(sorted(public_headers))
+    target_dict['headers'] = list(sorted(headers))
+    target_dict['src'] = list(sorted(src))
+    target_dict['deps'] = list(sorted(deps))
+
+
+def _generate_build_metadata(build_extra_metadata, bazel_rules):
+    lib_names = build_extra_metadata.keys()
+    result = {}
+    
+    for lib_name in lib_names:
+        lib_dict = _create_target_from_bazel_rule(lib_name, bazel_rules)
+
+        _expand_intermediate_deps(lib_dict, lib_names, bazel_rules)    
+
+        # populate extra properties from build metadata
+        lib_dict.update(build_extra_metadata.get(lib_name, {}))
+
+        # store to results
+        result[lib_name] = lib_dict
+
+    # rename some targets to something else
+    # this needs to be made after we're done with most of processing logic
+    # otherwise the already-renamed libraries will have different names than expected
+    for lib_name in lib_names:
+        to_name = build_extra_metadata.get(lib_name, {}).get('_RENAME', None)
+        if to_name:
+            # store lib under the new name and also change its 'name' property
+            if to_name in result:
+                raise Exception('Cannot rename target ' + lib_name + ', ' + to_name + ' already exists.')
+            lib_dict = result.pop(lib_name)
+            lib_dict['name'] = to_name
+            result[to_name] = lib_dict
+
+            # dep names need to be updated as well
+            for lib_dict_to_update in result.values():
+                lib_dict_to_update['deps'] = list(map(lambda dep: to_name if dep == lib_name else dep, lib_dict_to_update['deps']))
+
+    # make sure deps are listed in reverse topological order (e.g. "grpc gpr" and not "gpr grpc")
+    for lib_dict in result.itervalues():
+        lib_dict['deps'] = list(reversed(_sort_by_build_order(lib_dict['deps'], result)))
+
+    return result
+
+
+def _convert_to_build_yaml_like(lib_dict):
+    lib_names = list(filter(lambda lib_name: lib_dict[lib_name].get('_TYPE', 'library') == 'library' , lib_dict.keys()))
+    target_names = list(filter(lambda lib_name: lib_dict[lib_name].get('_TYPE', 'library') == 'target' , lib_dict.keys()))
+    test_names = list(filter(lambda lib_name: lib_dict[lib_name].get('_TYPE', 'library') == 'test' , lib_dict.keys()))
+
+    # make sure libraries come in build order (seems to be required by Makefile)
+    lib_names = _sort_by_build_order(lib_names, lib_dict)
```

Comment:
This can be off-loaded by `transitive_dependencies.py` and previously `build.yaml` was already sorted alphabetically so it's a bit weird to make this sorted.

Reply:
I'll remove this sorting and makes sure build_cleaner.py sorts the libs, targets and tests in the same way they were sorted before.